keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/tensor/TensorNUTS.java

```diff
@@ -0,0 +1,531 @@
+        return random.nextDouble() < probability;
+    }
+
+    private static boolean isNotUTurning(Map<Long, DoubleTensor> positionForward,
+                                         Map<Long, DoubleTensor> positionBackward,
+                                         Map<Long, DoubleTensor> momentumForward,
+                                         Map<Long, DoubleTensor> momentumBackward) {
+        double forward = 0.0;
+        double backward = 0.0;
+
+        for (Map.Entry<Long, DoubleTensor> forwardPositionForLatent : positionForward.entrySet()) {
+
+            final long latentId = forwardPositionForLatent.getKey();
+            final DoubleTensor forwardMinusBackward = forwardPositionForLatent.getValue().minus(
+                positionBackward.get(latentId)
+            );
+
+            forward += forwardMinusBackward.times(momentumForward.get(latentId)).sum();
+            backward += forwardMinusBackward.times(momentumBackward.get(latentId)).sum();
+        }
+
+        return (forward >= 0.0) && (backward >= 0.0);
+    }
+
+    private static void cachePosition(List<Vertex<DoubleTensor>> latentVertices, Map<Long, DoubleTensor> position) {
+        for (Vertex<DoubleTensor> vertex : latentVertices) {
+            position.put(vertex.getId(), vertex.getValue());
+        }
+    }
+
+    private static void initializeMomentumForEachVertex(List<Vertex<DoubleTensor>> vertices,
+                                                        Map<Long, DoubleTensor> momentums,
+                                                        KeanuRandom random) {
+        for (Vertex<DoubleTensor> vertex : vertices) {
+            momentums.put(vertex.getId(), random.nextGaussian(vertex.getValue().getShape()));
+        }
+    }
+
+    private static void cache(Map<Long, DoubleTensor> from, Map<Long, DoubleTensor> to) {
+        for (Map.Entry<Long, DoubleTensor> entry : from.entrySet()) {
+            to.put(entry.getKey(), entry.getValue());
+        }
+    }
+
+    private static LeapFrogged leapfrog(final List<Vertex<DoubleTensor>> latentVertices,
+                                        final Map<Long, Long> latentSetAndCascadeCache,
+                                        final List<Vertex> probabilisticVertices,
+                                        final Map<Long, DoubleTensor> position,
+                                        final Map<Long, DoubleTensor> gradient,
+                                        final Map<Long, DoubleTensor> momentum,
+                                        final double epsilon) {
+
+        final double halfTimeStep = epsilon / 2.0;
+
+        Map<Long, DoubleTensor> nextMomentum = new HashMap<>();
+        Map<Long, DoubleTensor> nextPosition = new HashMap<>();
+
+        for (Map.Entry<Long, DoubleTensor> rEntry : momentum.entrySet()) {
+            final DoubleTensor updatedMomentum = rEntry.getValue().plus(gradient.get(rEntry.getKey()).times(halfTimeStep));
```

Comment:
plus(...).timesInPlace(halfTimeStep) ?

Reply:
That doesn't work unfortunately, because it's plus(gradient times halfTimeStep)

Reply:
Looks like we can reorder it to be: `final DoubleTensor updatedMomentum = (gradient.get(rEntry.getKey()).times(halfTimeStep)).plusInPlace(rEntry.getValue());`